src/conf_helpers.c

```diff
@@ -0,0 +1,92 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>  
+#include <signal.h>
+
+#include "mosquitto_broker_internal.h"
+
+#include "conf_helpers.h"
+
+
+int update_conf_param(struct mosquitto__config *config, const char *param_key, const char *param_value) {
+    if(strcmp(param_key, "allow_anonymous") == 0) {
+        if(strcmp(param_value, "true") == 0) {
+            config->security_options.allow_anonymous = -1;
+            return 0;
+        }
+        else if(strcmp(param_value, "false") == 0) {
+            config->security_options.allow_anonymous = 0;
+            return 0;
+        }
+    }
+
+    mosquitto_log_printf(MOSQ_LOG_WARNING, "Unrecognized configuration value!");
+    
+    return 1;
+}
+
+void update_allow_anonymous_conf(struct mosquitto__config *config, bool allow) {
+    if(config->per_listener_settings) {
+        for(int i = 0; i < config->listener_count; i++){
+            if(strcmp(config->listeners[i].host, "127.0.0.1") != 0 && strcmp(config->listeners[i].host, "localhost") != 0) {
+                config->listeners[i].security_options.allow_anonymous = allow;
+            }
+        }
+    }
+    else {
+        config->security_options.allow_anonymous = allow;
+    }
+}
+
+int write_conf_to_file(struct mosquitto__config *config, const char *file_path) {
+    FILE *file = fopen(file_path, "r");
+    if(file == NULL) {
+        mosquitto_log_printf(MOSQ_LOG_ERR, "Could not open config file!");
+        return 1;
+    }
+
+    FILE *temp_file = fopen("temp_config.conf", "w");
+    if(temp_file == NULL) {
+        fclose(file);
+        mosquitto_log_printf(MOSQ_LOG_ERR, "Could not create temporary file!");
+        return 1;
+    }
+
+    char line[256];
+
+    const char *new_global_line = config->security_options.allow_anonymous ? "allow_anonymous true\n" : "allow_anonymous false\n";
+
+    while(fgets(line, sizeof(line), file)) {
+        if(strstr(line, "listener") != NULL && strstr(line, "127.0.0.1") != NULL) {
+            fputs(line, temp_file);
+            if(fgets(line, sizeof(line), file)) {
```

Comment:
So here you meet `listener` and `127.0.0.1` and then read the next line and don't check it... what if the next line also has `listener` and `127.0.0.1` followed by allow_anonymous?

Reply:
To not change the settings under local listener, I just write them without even checking. If there is another listener under local listener, it finds and edits its allow_anonymous value accordingly with the outside while.

Reply:
There can be many settings under a listener that relate to that listener, so why do you always assume that there is only one setting? And btw, there can be also blank lines...

And in cases like this one, it will change allow_anonymous while it must not:
```
listener 8883 127.0.0.1
listener 8885 127.0.0.1
allow_annonymous = true
```

Reply:
I don't assume that there will be only one setting, it will write every setting as it is when it saw a local listener. I am adding output from my trial here. As far as I try it is not properly working only when there is two allow_anonymous line in one local listener.

![after](https://github.com/user-attachments/assets/9a0a9ffb-2bb5-4741-9c1b-709cb6f4a7c0)

![before](https://github.com/user-attachments/assets/03ab3ace-977e-4bcd-b07d-7157b719e69c)

Reply:
No, it's not working, I've just tried myself running your changes;
This is what I call:
```
	update_allow_anonymous(db.config, true);
	config__write(db.config, db.config_file);
```
And here is a result. As you can see both `allow_anonymous` are changed.

Before:
```
kat@desktop:$ cat mosquitto-local.conf
plugin /all/mine/pluraf/iotplan/repos/m-broker-c/build/plugins/dynamic-security/mosquitto_dynamic_security.so
plugin_opt_config_file /all/mine/pluraf/iotplan/repos/m-broker-c/configs/dynamic-security.json

http_api_pkey_file /all/mine/pluraf/iotplan/repos/m-broker-c/bits/jwt_public_key.pem

allow_anonymous false

log_type all

listener 8883 0.0.0.0
tls_version tlsv1.2
certfile /all/mine/pluraf/iotplan/repos/m-broker-c/tls/mqbc.crt
keyfile /all/mine/pluraf/iotplan/repos/m-broker-c/tls/mqbc.key

listener 1883 0.0.0.0

listener 1884 127.0.0.1
listener 1885 127.0.0.1
allow_anonymous false


listener 1880 127.0.0.1

```

After:
```
kat@desktop:$ cat mosquitto-local.conf
plugin /all/mine/pluraf/iotplan/repos/m-broker-c/build/plugins/dynamic-security/mosquitto_dynamic_security.so
plugin_opt_config_file /all/mine/pluraf/iotplan/repos/m-broker-c/configs/dynamic-security.json

http_api_pkey_file /all/mine/pluraf/iotplan/repos/m-broker-c/bits/jwt_public_key.pem

allow_anonymous true

log_type all

listener 8883 0.0.0.0
tls_version tlsv1.2
certfile /all/mine/pluraf/iotplan/repos/m-broker-c/tls/mqbc.crt
keyfile /all/mine/pluraf/iotplan/repos/m-broker-c/tls/mqbc.key

listener 1883 0.0.0.0

listener 1884 127.0.0.1
listener 1885 127.0.0.1
allow_anonymous true


listener 1880 127.0.0.1

```

Reply:
There was a logical error where I handled excluding localhost, I believe that is why we have different results. I fixed it. But I am working on fixing the issue related with having allow_anonymous under other settings.

Reply:
I believe the issue is fixed.
before:
![before](https://github.com/user-attachments/assets/99a3e5b1-1fb0-4194-b77f-8b772251ca30)
after:
![after](https://github.com/user-attachments/assets/46482ec9-fbb0-4cac-bcdc-d3c0b41fa569)